lib/cfme/cloud_services/payload_builder.rb

```diff
@@ -0,0 +1,86 @@
+require 'minitar'
+
+module Cfme
+  module CloudServices
+    module PayloadBuilder
+      BOUNDARY = "AaB03x".freeze
+
+      module RequestParams
+        UPLOADER_HOST = "localhost".freeze
+        UPLOADER_PORT = "8080".freeze
+        UPLOADER_URI = "/api/ingress/v1/upload".freeze
+
+        CONTENT_TYPE = "application/vnd.redhat.topological-inventory.something+tgz".freeze
+
+        def org_id
+          54_321 # TODO: tie with subscription
+        end
+
+        def tenant
+          12_345 # TODO: tie with subscription
+        end
+
+        def x_rh_insights_request_id
+          # TODO: generate?
+          # request_id: Unique ID provided to the payload created by 3Scale.
+          # This ID will be used for the life of the object.
+          "52df9f748eabcfea"
+        end
+
+        def identity_content
+          {"identity" => {"account_number" => tenant}, "internal" => {"org_id" => org_id}}
+        end
+
+        def identity_headers
+          {"x-rh-identity" => Base64.strict_encode64(JSON.dump(identity_content)) }
+        end
+
+        def header
+          {"Content-Type"          => "multipart/form-data;boundary=\"#{BOUNDARY}\"",
+           "x-rh-insights-request" => x_rh_insights_request_id}.merge(identity_headers)
+        end
+
+        def inventory_file_name
+          "cfme_invetory_#{tenant}_#{org_id}"
+        end
+      end
+
+      def create_tempfile(json_data, source)
+        file = Tempfile.new(source)
+        file.write(json_data)
+        file.rewind
+        file
+      end
+
+      def gzipped_tar_content_from(payload)
+        begin
+          tempfile = create_tempfile(JSON.generate(payload), inventory_file_name)
+          file = StringIO.new(String.new)
+          sgz = ::Zlib::GzipWriter.new(file)
+          tar = ::Minitar::Output.new(sgz)
+
+          ::Minitar.pack_file(tempfile.path, tar)
```

Comment:
I'm good with this for now...I think @agrare had a way to IO stream this, but we can always change it.

Reply:
Yeah rubygems has builtin tar handling so you don't need an extra gem, my code isn't anywhere where it can be reused by this (it is in [topological_inventory-sync](https://github.com/ManageIQ/topological_inventory-sync/blob/master/lib/topological_inventory/sync/inventory_upload/parser.rb#L21-L28))

Reply:
done 👍